servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/ClientIpUtils.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.http.api.HttpHeaderNames;
+import io.servicetalk.http.api.HttpHeaders;
+import io.servicetalk.transport.api.ConnectionContext;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.http.api.HttpHeaderNames.FORWARDED;
+import static io.servicetalk.http.api.HttpHeaderNames.X_FORWARDED_FOR;
+
+/**
+ * Client helper functions to extract actual client IP from headers when request goes
+ * through a forwarding proxy/proxies.
+ * Worth noting that these headers can be altered by the originator, so additional validation may be required
+ * to validate the origin.
+ */
+public final class ClientIpUtils {
+
+    private ClientIpUtils() {
+        // no instances
+    }
+
+    /**
+     * Extracts the client IP address from request headers or connection context.
+     *
+     * @param headers an {@link HttpHeaders} of the request
+     * @param ctx a {@link ConnectionContext} of the request
+     * @param fallbackToSocketAddress whether to fallback to the remote address in {@link ConnectionContext} if no
+     * client IP header is present
+     * @return inferred client IP address as {@link CharSequence}
+     */
+    @Nullable
+    static CharSequence extractClientIp(final HttpHeaders headers,
+                                        final ConnectionContext ctx,
+                                        final boolean fallbackToSocketAddress) {
+
+        final CharSequence ip = parseXForwardedFor(headers.get(X_FORWARDED_FOR));
+        if (ip != null) {
+            return ip;
+        }
+
+        final CharSequence ipFromForwarded = parseForwarded(headers.get(FORWARDED));
```

Comment:
Shouldn't the `Forwarded` header be considered first if it exists as it's a standardized header (https://datatracker.ietf.org/doc/html/rfc7239) and only if it's missing should the `X-Forwarded-For` be considered?
Also, the RFC recommends mapping `XFF` to `Forwarded` and have a unified parsing strategy: https://datatracker.ietf.org/doc/html/rfc7239#section-7.4